be/src/storage/persistent_index.cpp

```diff
@@ -416,52 +423,93 @@ StatusOr<std::unique_ptr<ImmutableIndexShard>> ImmutableIndexShard::create(size_
     return std::move(ret);
 }
 
-Status write_immutable_index(size_t key_size, size_t value_size,
-                             const std::vector<std::vector<IndexHash>>& hashes_by_shard,
-                             const std::vector<std::vector<KVPairPtr>>& kv_ptrs_by_shard, size_t npage_hint,
-                             const EditVersion& version, fs::WritableBlock& wb) {
-    size_t kv_size = key_size + value_size;
-    ImmutableIndexMetaPB meta;
-    size_t total = 0;
-    size_t total_moved = 0;
-    size_t total_bytes = 0;
-    for (size_t i = 0; i < hashes_by_shard.size(); i++) {
-        auto rs_create = ImmutableIndexShard::create(kv_size, hashes_by_shard[i], kv_ptrs_by_shard[i], npage_hint);
+class ImmutableIndexWriter {
+public:
+    ~ImmutableIndexWriter() {
+        if (_wb) {
+            Env::Default()->delete_file(_idx_file_path_tmp);
+        }
+    }
+
+    Status init(const string& dir, const EditVersion& version) {
+        _version = version;
+        _idx_file_path = strings::Substitute("$0/index.l1.$1.$2", dir, version.major(), version.minor());
+        _idx_file_path_tmp = _idx_file_path + ".tmp";
+        fs::BlockManager* block_mgr = fs::fs_util::block_manager();
+        fs::CreateBlockOptions wblock_opts({_idx_file_path_tmp});
```

Comment:
The default mode of CreateBlockOptions is `MUST_CREATE`. If be crash during merge compaction and tmp file is not deleted, the next merge compaction maybe failed.